arch/arm/core/aarch32/thread.c

```diff
@@ -420,3 +430,54 @@ void arch_switch_to_main_thread(struct k_thread *main_thread, char *stack_ptr,
 
 	CODE_UNREACHABLE;
 }
+
+#if !defined(CONFIG_MULTITHREADING) && defined(CONFIG_CPU_CORTEX_M)
+
+extern K_THREAD_STACK_DEFINE(z_main_stack, CONFIG_MAIN_STACK_SIZE);
+
+FUNC_NORETURN void z_arm_switch_to_main_no_multithreading(
+	k_thread_entry_t main_entry, void *p1, void *p2, void *p3)
+{
+	z_arm_prepare_switch_to_main();
+
+	/* Set PSP to the highest address of the main stack. */
+	char *psp =	Z_THREAD_STACK_BUFFER(z_main_stack) +
+		K_THREAD_STACK_SIZEOF(z_main_stack);
+
+#if defined(CONFIG_BUILTIN_STACK_GUARD)
+	char *psplim = (Z_THREAD_STACK_BUFFER(z_main_stack));
+	/* Clear PSPLIM before setting it to guard the main stack area. */
+	__set_PSPLIM(0);
+#endif
+
+	__asm__ volatile (
+	"mov  r0, %0\n\t"
+	"mov  r1, %1\n\t"
+	"mov  r2, %2\n\t"
+	"mov  r4, %3\n\t"
+	"mov  r5, %5\n\t"
```

Comment:
What if those registers get allocated for inputs `%4` or `%6`? Or `r4` for input `%5`?

Reply:
Yes, these need to be clobbered

Reply:
And that's also a reason why I think using local register variables is a better approach. You don't get into such problems, as the compiler cares about used registers itself.

Reply:
Alright, I 'll switch to using local register values.
I have actually taken this approach in the system calls generation code.

Reply:
Updated: I defined local register variables, to hold the required input, after the stack pointer has changed.
I had to use high registers, as the total constraints' set for Cortex-M0 was resulting to a non-feasible implementation.

Now I  still need to use manual "mov" instructions to copy the arguments to r0,r1,r2, since i don't clobber these registers any more. That's for extra safety.